desdeo/mcdm/pareto_navigator.py

```diff
@@ -0,0 +1,271 @@
+
+    for i in range(len(pref_info["classification"])):
+        if pref_info["classification"][i] == "<":
+            ref.append(ideal[i])
+        elif pref_info["classification"][i] == ">":
+            ref.append(nadir[i])
+        elif pref_info["classification"][i] == "=":
+            ref.append(sol[i])
+    return numpy_array_to_objective_dict(problem, np.array(ref))
+
+def calculate_adjusted_speed(allowed_speeds: np.ndarray, speed: float) -> float:
+    """Calculate an adjusted speed from a given float.
+
+    Note:
+        Adjusting the speed is not specified in the article but seems necessary.
+
+    Args:
+        allowed_speeds (np.ndarray): An array of allowed speeds.
+        speed (int): A given speed value where.
+
+    Returns:
+        float: An adjusted speed value calculated from given float.
+            Is between 0 and 1.
+    """
+    return (speed / np.max(allowed_speeds)) / 20
+
+def calculate_search_direction(
+    problem: Problem,
+    reference_point: dict[str, float],
+    current_point: dict[str, float]
+) -> dict[str, float]:
+    """Calculate search direction from the current point to the reference point.
+
+    Args:
+        problem (Problem): The problem being solved.
+        reference_point (dict[str, float]): The given reference point.
+        current_point (dict[str, float]): Currently navigated point.
+
+    Returns:
+        dict[str, float]: The direction from the current point to the reference point.
+    """
+    z = objective_dict_to_numpy_array(problem, current_point)
+    q = objective_dict_to_numpy_array(problem, reference_point)
+
+    d = q - z
+    return numpy_array_to_objective_dict(problem, d)
+
+def get_polyhedral_set(problem: Problem) -> tuple[np.ndarray, np.ndarray]:
+    """Get a polyhedral set as convex hull from the set of pareto optimal solutions.
+
+    Args:
+        problem (Problem): The problem being solved.
+
+    Returns:
+        tuple[np.ndarray, np.ndarray]: The A matrix and b vector from the polyhedral set equation.
+    """
+    # should add checking if dicrete_representation doesn't exist and handle it
+    objective_values = problem.discrete_representation.objective_values
+    representation = np.array([objective_values[obj.symbol] for obj in problem.objectives])
+
+    convex_hull = ConvexHull(representation.T)
+    matrix_a = convex_hull.equations[:, 0:-1]
+    b = -convex_hull.equations[:, -1]
+    return matrix_a, b
+
+def construct_matrix_a(problem: Problem, matrix_a: np.ndarray) -> np.ndarray:
+    """Construct the A' matrix in the linear parametric programming problem from the article.
+
+    Args:
+        problem (Problem): The problem being solved.
+        matrix_a (np.ndarray): The A matrix from the polyhedral set equation.
+
+    Returns:
+        np.ndarray: The A' matrix in the linear parametric programming problem from the article.
+    """
+    ideal = objective_dict_to_numpy_array(problem, get_ideal_dict(problem))
```

Comment:
See my previous comment. We have methods now for the `Problem` class that return the ideal and nadir points as dicts. Consider refactoring them here.

Reply:
Vaihdoin nuo get_ideal_dict() ja nadirin vastaavat Problem.get_ideal_point() muotoon koko tiedostoon.